### PR TITLE
Match NumPy by considering NaNs to be larger than any number when sorting

### LINKED ISSUE #15764: torch.argsort descends wrongly
## 🐛 Bug
given a large amount of numbers (say, a list containing 32768 numbers), I found that the argsort behaves wrongly when setting the descend=False.
## To Reproduce
I want to calculate the similarity of word embeddings, and I managed to achieve that by batched-similarity calculation. I expect the most similarity to an embedding should be itself, thus the highest rank should be 1,2... accordingly, and vice versa. 
now say we have torch.tensor object `similarity` (in shape batch_size * 32768), I want to extract top similarities along the last dimension.
```
get_top = torch.argsort(similarity, dim=1)
get_top = torch.argsort(similarity, dim=1, descending=True) 
```
the code gives the result of:

> get_top = torch.argsort(similarity, dim=1)
> [20485, 30807, 27706,  ...,   117,     6,     **1**],
>  [29628, 16835, 23989,  ...,  6773, 10377,     **2**],
>    ...,
>  [30120, 23853, 24914,  ...,   118,    50,    **47**],
>  [23528, 23663, 25283,  ...,    78,    77,    **48**],
>  [31738, 17062, 29731,  ...,  1240,    79,    **49**]

while:

> top8 = torch.argsort(similarity, dim=1, descending=True)
> tensor([[**21820**, 21822, 21816,  ..., 10909, 10910, 32725],
>         [**32497**, 28158, 20713,  ..., 14889, 12632, 16177],
>         [**17110**, 17460, 18861,  ..., 14628, 12113,    18],
>         ...,
>         [**22335**, 21440, 26360,  ..., 14870,  8292, 13842],
>         [**20379**, 16479, 28973,  ..., 11770, 11327, 13454],
>         [**23388**, 19492, 16577,  ...,  7874,  5219, 13230]])
the bold part implys the main problems. and I assume it's not simply because of the accuracy of float numebrs.

## Environment
 - PyTorch Version: 1.0 stable
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip install
 - Python version: 3.6.3

### PR DESCRIPTION
**Pull request by umanwizard**
Fixes #15764

### CODE REVIEW

**Comment by ailzhang** [comment]
Thanks for working on this! With `sort` methods I think we also have problems with `float('-inf')`. Is it going to be in a separate PR btw?

**Comment by umanwizard** [comment]
@ailzhang What is the problem with `-inf` ? It seems to work as I expect:

```
In [3]: x
Out[3]: tensor([3., -inf, 1.])

In [4]: x.sort()
Out[4]: (tensor([-inf, 1., 3.]), tensor([1, 2, 0]))

In [5]: x.sort(descending=True)
Out[5]: (tensor([3., 1., -inf]), tensor([0, 2, 1]))
```

**Comment by ailzhang** [comment]
Actually the problem was indeed caused by `nan`. `ninf` is sorted correctly. Please ignore my comment then. Thanks @umanwizard !

**Review by ailzhang — APPROVED** [review]
LGTM! CI failures seem unrelated except the clang-tidy one. Good to go once CI is green.

**Comment by ngimel** [comment]
Both thrust and quicksort in TH require strict weak ordering comparator https://thrust.github.io/doc/group__sorting.html#ga3f47925d80f4970d5730051dba1c5603. 
```
#define GT_OR_NAN(x, y) \
  ((x != x) || (x > y))
```
is not strict weak order.

**Comment by ailzhang** [comment]
@ngimel yea I agree on the solution is not strict weak order. Hmmm why do we enforce strict weak ordering btw (could you give an example) ? Thanks!

**Comment by umanwizard** [comment]
@ngimel Thank you for the hint. If I understand correctly, I can fix this by making those macros return false if both the arguments are NaN.
